### ISSUE TITLE
When nodes are recreated with bootstrapping on GCP (and others?) nodes lose SDN network configuration

### ISSUE THREAD

**Issue by smarterclayton** [openshift/origin#17125]
1. Create a cluster using golden images on GCP, with nodes in an autoscaling managed group
2. Create pods on both masters and nodes:

```
$ oc get pods --all-namespaces -o wide
NAMESPACE         NAME                       READY     STATUS    RESTARTS   AGE       IP           NODE
default           docker-registry-1-46jkg    1/1       Running   0          22m       172.16.6.5   ci-claytontest3-ig-m-3c6l
default           registry-console-1-6qwc9   1/1       Running   0          21m       172.16.4.2   ci-claytontest3-ig-n-vcqc
default           router-1-msspw             1/1       Running   0          21m       10.142.0.2   ci-claytontest3-ig-m-3c6l
kube-system       prometheus-0               5/5       Running   0          13m       172.16.2.3   ci-claytontest3-ig-n-7zsf
openshift-infra   bootstrap-autoapprover-0   1/1       Running   0          22m       172.16.6.2   ci-claytontest3-ig-m-3c6l
$ oc get hostsubnets
NAME                        HOST                        HOST IP      SUBNET          EGRESS IPS
ci-claytontest3-ig-m-3c6l   ci-claytontest3-ig-m-3c6l   10.142.0.2   172.16.6.0/23   []
ci-claytontest3-ig-n-7zsf   ci-claytontest3-ig-n-7zsf   10.142.0.3   172.16.2.0/23   []
ci-claytontest3-ig-n-svf9   ci-claytontest3-ig-n-svf9   10.142.0.5   172.16.0.0/23   []
ci-claytontest3-ig-n-vcqc   ci-claytontest3-ig-n-vcqc   10.142.0.4   172.16.4.0/23   []
```

3. Delete the nodes from GCP (which triggers a node object deletion, and then new instances are recreated with the same names)
4. Observe that nodes get new subnets

```
$ oc get hostsubnets
NAME                        HOST                        HOST IP      SUBNET           EGRESS IPS
ci-claytontest3-ig-m-3c6l   ci-claytontest3-ig-m-3c6l   10.142.0.2   172.16.6.0/23    []
ci-claytontest3-ig-n-7zsf   ci-claytontest3-ig-n-7zsf   10.142.0.3   172.16.8.0/23    []
ci-claytontest3-ig-n-svf9   ci-claytontest3-ig-n-svf9   10.142.0.5   172.16.10.0/23   []
ci-claytontest3-ig-n-vcqc   ci-claytontest3-ig-n-vcqc   10.142.0.4   172.16.12.0/23   []
```

BUG: Pods that were previously assigned to those nodes get recreated, BUT they have the old IPs (on old subnets), and are not given new IPs.  The pods are unreachable until deleted and recreated.  However, because this is a new node the pods should have been allocated new IPs on the new subnet, not the old one.

```
$ oc get pods --all-namespaces -o wide
NAMESPACE         NAME                       READY     STATUS    RESTARTS   AGE       IP           NODE
default           docker-registry-1-46jkg    1/1       Running   0          42m       172.16.6.5   ci-claytontest3-ig-m-3c6l
default           registry-console-1-nf9c4   1/1       Running   0          10m       172.16.6.6   ci-claytontest3-ig-m-3c6l
default           router-1-msspw             1/1       Running   0          42m       10.142.0.2   ci-claytontest3-ig-m-3c6l
kube-system       prometheus-0               5/5       Running   0          34m       172.16.2.2   ci-claytontest3-ig-n-7zsf
openshift-infra   bootstrap-autoapprover-0   1/1       Running   1          43m       172.16.6.2   ci-claytontest3-ig-m-3c6l
```

This can probably happen even on an existing OpenShift cluster - if you stop an instance and delete the node object, and on start it comes back up and announces the same identity, it's likely this will happen.  It's just easier to trigger now with bootstrapping and golden images.

I can provide a cluster to debug against.  @openshift/sig-networking

**Comment by danwinship** [comment]
> BUG 1: Master gets a new subnet, it shouldn't because it didn't change

I think you misread? Assuming the one with `-m-` in the name is the master, it has `172.16.6.0/23` both before and after

**Comment by smarterclayton** [comment]
You're right, too many 6's and 4's and 2's

**Comment by danwinship** [comment]
> This can probably happen even on an existing OpenShift cluster - if you stop an instance and delete the node object, and on start it comes back up and announces the same identity, it's likely this will happen.

Yes, *if* the pod's containers stay running. Eg, if I run a pod on a node, stop atomic-openshift-node, delete its Node object, and then start atomic-openshift-node again, then the node will get assigned a new HostSubnet but the pod will keep running with its old IP, as above. (Which is kinda weird... Shouldn't the master mark the Pod as Failed if its node gets deleted?)

But if I throw in a "systemctl restart docker" before starting the node service again, so that the pod's existing containers get killed, then when the pod containers are re-created, they get a new IP on the new HostSubnet.

Are you sure the pod is actually getting restarted even? (The pod with the bad IP in the output above lists "0" restarts.) Maybe something is going wrong that's keeping the master from figuring out that the pod is dead, and so the problem isn't "it's unreachable because it has a bad IP" but rather "it's unreachable because it's not actually running".

**Comment by smarterclayton** [comment]
The nodes and their disks are deleted.  Another theory is that it happens
fast enough that the node controller doesn’t evict those pods.  I can set
up a recreate tomorrow

On Nov 2, 2017, at 1:34 PM, Dan Winship <notifications@github.com> wrote:

This can probably happen even on an existing OpenShift cluster - if you
stop an instance and delete the node object, and on start it comes back up
and announces the same identity, it's likely this will happen.

Yes, *if* the pod's containers stay running. Eg, if I run a pod on a node,
stop atomic-openshift-node, delete its Node object, and then start
atomic-openshift-node again, then the node will get assigned a new
HostSubnet but the pod will keep running with its old IP, as above. (Which
is kinda weird... Shouldn't the master mark the Pod as Failed if its node
gets deleted?)

But if I throw in a "systemctl restart docker" before starting the node
service again, so that the pod's existing containers get killed, then when
the pod containers are re-created, they get a new IP on the new HostSubnet.

Are you sure the pod is actually getting restarted even? (The pod with the
bad IP in the output above lists "0" restarts.) Maybe something is going
wrong that's keeping the master from figuring out that the pod is dead, and
so the problem isn't "it's unreachable because it has a bad IP" but rather
"it's unreachable because it's not actually running".

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/issues/17125#issuecomment-341499830>,
or mute the thread

**Comment by smarterclayton** [comment]
Another possibility - sdnncontroller notices the node is dead and deletes
and recreates the hostsubnet.  But the new node sees the old version of the
subnet before it is refreshed.

On Nov 2, 2017, at 1:34 PM, Dan Winship <notifications@github.com> wrote:

This can probably happen even on an existing OpenShift cluster - if you
stop an instance and delete the node object, and on start it comes back up
and announces the same identity, it's likely this will happen.

Yes, *if* the pod's containers stay running. Eg, if I run a pod on a node,
stop atomic-openshift-node, delete its Node object, and then start
atomic-openshift-node again, then the node will get assigned a new
HostSubnet but the pod will keep running with its old IP, as above. (Which
is kinda weird... Shouldn't the master mark the Pod as Failed if its node
gets deleted?)

But if I throw in a "systemctl restart docker" before starting the node
service again, so that the pod's existing containers get killed, then when
the pod containers are re-created, they get a new IP on the new HostSubnet.

Are you sure the pod is actually getting restarted even? (The pod with the
bad IP in the output above lists "0" restarts.) Maybe something is going
wrong that's keeping the master from figuring out that the pod is dead, and
so the problem isn't "it's unreachable because it has a bad IP" but rather
"it's unreachable because it's not actually running".

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/issues/17125#issuecomment-341499830>,
or mute the thread

**Comment by smarterclayton** [comment]
Fairly easy to recreate:

```
# existing node
apiVersion: v1
host: ci-claytontest-ig-n-d363
hostIP: 10.142.0.5
kind: HostSubnet
metadata:
  creationTimestamp: 2017-11-03T23:04:52Z
  name: ci-claytontest-ig-n-d363
  resourceVersion: "1429"
  selfLink: /oapi/v1/hostsubnets/ci-claytontest-ig-n-d363
  uid: 6667319b-c0eb-11e7-b34d-42010a8e0004
subnet: 172.16.6.0/23
apiVersion: v1
host: ci-claytontest-ig-n-l1t9
hostIP: 10.142.0.3
kind: HostSubnet
metadata:
  creationTimestamp: 2017-11-03T23:04:52Z
  name: ci-claytontest-ig-n-l1t9
  resourceVersion: "1417"
  selfLink: /oapi/v1/hostsubnets/ci-claytontest-ig-n-l1t9
  uid: 663a67d1-c0eb-11e7-b34d-42010a8e0004
subnet: 172.16.4.0/23

# trigger delete

# event DELETE of hostsubnet
apiVersion: v1
host: ci-claytontest-ig-n-d363
hostIP: 10.142.0.5
kind: HostSubnet
metadata:
  creationTimestamp: 2017-11-03T23:04:52Z
  name: ci-claytontest-ig-n-d363
  resourceVersion: "1429"
  selfLink: /oapi/v1/hostsubnets/ci-claytontest-ig-n-d363
  uid: 6667319b-c0eb-11e7-b34d-42010a8e0004
subnet: 172.16.6.0/23

# event CREATE of new hostsubnet
apiVersion: v1
host: ci-claytontest-ig-n-d363
hostIP: 10.142.0.5
kind: HostSubnet
metadata:
  creationTimestamp: 2017-11-03T23:31:07Z
  name: ci-claytontest-ig-n-d363
  resourceVersion: "4335"
  selfLink: /oapi/v1/hostsubnets/ci-claytontest-ig-n-d363
  uid: 113cec23-c0ef-11e7-b34d-42010a8e0004
subnet: 172.16.8.0/23
```

Repeated, watching the node this time:

```
# last event on old pod, DELETE by node controller
apiVersion: v1
kind: Node
metadata:
  annotations:
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: 2017-11-03T23:31:07Z
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: n1-standard-2
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-east1
    failure-domain.beta.kubernetes.io/zone: us-east1-c
    kubernetes.io/hostname: ci-claytontest-ig-n-d363
    role: app
  name: ci-claytontest-ig-n-d363
  resourceVersion: "4732"
  selfLink: /api/v1/nodes/ci-claytontest-ig-n-d363
  uid: 113b3610-c0ef-11e7-b34d-42010a8e0004
spec:
  externalID: "4670166334054679283"
  providerID: gce://openshift-gce-devel-ci/us-east1-c/ci-claytontest-ig-n-d363
status:
  addresses:
  - address: 10.142.0.5
    type: InternalIP
  - address: 35.185.67.97
    type: ExternalIP
  - address: ci-claytontest-ig-n-d363
    type: Hostname
  allocatable:
    cpu: "1"
    memory: 6392080Ki
    pods: "20"
  capacity:
    cpu: "2"
    memory: 7494480Ki
    pods: "20"
  conditions:
  - lastHeartbeatTime: null
    lastTransitionTime: 2017-11-03T23:31:07Z
    message: openshift-sdn cleared kubelet-set NoRouteCreated
    reason: RouteCreated
    status: "False"
    type: NetworkUnavailable
  - lastHeartbeatTime: 2017-11-03T23:34:17Z
    lastTransitionTime: 2017-11-03T23:34:58Z
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: OutOfDisk
  - lastHeartbeatTime: 2017-11-03T23:34:17Z
    lastTransitionTime: 2017-11-03T23:34:58Z
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: MemoryPressure
  - lastHeartbeatTime: 2017-11-03T23:34:17Z
    lastTransitionTime: 2017-11-03T23:34:58Z
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: DiskPressure
  - lastHeartbeatTime: 2017-11-03T23:34:17Z
    lastTransitionTime: 2017-11-03T23:34:58Z
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: Ready
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  images:
  - names:
    - docker.io/cockpit/kubernetes@sha256:a4b418f80f5654021ff63526bbb9644f23a5cc56b6eb50d684f21c1aaf324256
    - docker.io/cockpit/kubernetes:latest
    sizeBytes: 352215755
  - names:
    - docker.io/openshift/origin-pod@sha256:616b850e04084409d23d61a33fd367af7fae189b4ce45c037d3628b13197a524
    - docker.io/openshift/origin-pod:v3.7.0-rc.0
    sizeBytes: 218419873
  nodeInfo:
    architecture: amd64
    bootID: cbb28c23-cfe4-4f6d-a1a5-a5e0ae8f0c4c
    containerRuntimeVersion: docker://1.12.6
    kernelVersion: 3.10.0-693.5.2.el7.x86_64
    kubeProxyVersion: v1.7.6+a08f5eeb62
    kubeletVersion: v1.7.6+a08f5eeb62
    machineID: ded576e97e0cf2a27b2b930f82c77de0
    operatingSystem: linux
    osImage: Red Hat Enterprise Linux Server 7.4 (Maipo)
    systemUUID: A8705FD8-5E2C-3E97-4197-E5A12B1B45FF

# New pod registers itself
apiVersion: v1
kind: Node
metadata:
  annotations:
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: 2017-11-03T23:35:53Z
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: n1-standard-2
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-east1
    failure-domain.beta.kubernetes.io/zone: us-east1-c
    kubernetes.io/hostname: ci-claytontest-ig-n-d363
    role: app
  name: ci-claytontest-ig-n-d363
  resourceVersion: "4862"
  selfLink: /api/v1/nodes/ci-claytontest-ig-n-d363
  uid: bb8f2baa-c0ef-11e7-b34d-42010a8e0004
spec:
  externalID: "8737322233893045205"
  providerID: gce://openshift-gce-devel-ci/us-east1-c/ci-claytontest-ig-n-d363
status:
  addresses:
  - address: 10.142.0.5
    type: InternalIP
  - address: 35.185.115.5
    type: ExternalIP
  - address: ci-claytontest-ig-n-d363
    type: Hostname
  allocatable:
    cpu: "1"
    memory: 6392080Ki
    pods: "20"
  capacity:
    cpu: "2"
    memory: 7494480Ki
    pods: "20"
  conditions:
  - lastHeartbeatTime: null
    lastTransitionTime: 2017-11-03T23:35:53Z
    message: Node created without a route
    reason: NoRouteCreated
    status: "True"
    type: NetworkUnavailable
  - lastHeartbeatTime: 2017-11-03T23:35:53Z
    lastTransitionTime: 2017-11-03T23:35:53Z
    message: kubelet has sufficient disk space available
    reason: KubeletHasSufficientDisk
    status: "False"
    type: OutOfDisk
  - lastHeartbeatTime: 2017-11-03T23:35:53Z
    lastTransitionTime: 2017-11-03T23:35:53Z
    message: kubelet has sufficient memory available
    reason: KubeletHasSufficientMemory
    status: "False"
    type: MemoryPressure
  - lastHeartbeatTime: 2017-11-03T23:35:53Z
    lastTransitionTime: 2017-11-03T23:35:53Z
    message: kubelet has no disk pressure
    reason: KubeletHasNoDiskPressure
    status: "False"
    type: DiskPressure
  - lastHeartbeatTime: 2017-11-03T23:35:53Z
    lastTransitionTime: 2017-11-03T23:35:53Z
    message: 'container runtime is down,runtime network not ready: NetworkReady=false
      reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni
      config uninitialized'
    reason: KubeletNotReady
    status: "False"
    type: Ready
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  nodeInfo:
    architecture: amd64
    bootID: 55863715-fdcf-467f-9b7e-150c23c6f013
    containerRuntimeVersion: docker://1.12.6
    kernelVersion: 3.10.0-693.5.2.el7.x86_64
    kubeProxyVersion: v1.7.6+a08f5eeb62
    kubeletVersion: v1.7.6+a08f5eeb62
    machineID: ded576e97e0cf2a27b2b930f82c77de0
    operatingSystem: linux
    osImage: Red Hat Enterprise Linux Server 7.4 (Maipo)
    systemUUID: A8705FD8-5E2C-3E97-4197-E5A12B1B45FF
```

But when I tried this while watching a pod, I noticed that the pod was sometimes not deleted (the actual problem):

```
# existing pod 
{
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": {
        "annotations": {
            "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"registry-console-1\",\"uid\":\"4b81a94b-c0eb-11e7-b34d-42010a8e0004\",\"apiVersion\":\"v1\",\"resourceVersion\":\"5599\"}}\n",
            "openshift.io/deployment-config.latest-version": "1",
            "openshift.io/deployment-config.name": "registry-console",
            "openshift.io/deployment.name": "registry-console-1",
            "openshift.io/generated-by": "OpenShiftNewApp",
            "openshift.io/scc": "restricted"
        },
        "creationTimestamp": "2017-11-03T23:42:58Z",
        "generateName": "registry-console-1-",
        "labels": {
            "app": "registry-console",
            "deployment": "registry-console-1",
            "deploymentconfig": "registry-console",
            "name": "registry-console"
        },
        "name": "registry-console-1-lc9g6",
        "namespace": "default",
        "ownerReferences": [
            {
                "apiVersion": "v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "ReplicationController",
                "name": "registry-console-1",
                "uid": "4b81a94b-c0eb-11e7-b34d-42010a8e0004"
            }
        ],
        "resourceVersion": "5683",
        "selfLink": "/api/v1/namespaces/default/pods/registry-console-1-lc9g6",
        "uid": "b8e9cdc0-c0f0-11e7-b34d-42010a8e0004"
    },
    "spec": {
        "containers": [
            {
                "env": [
                    {
                        "name": "OPENSHIFT_OAUTH_PROVIDER_URL",
                        "value": "https://api.claytontest.origin-ci-int-gce.dev.rhcloud.com/"
                    },
                    {
                        "name": "OPENSHIFT_OAUTH_CLIENT_ID",
                        "value": "cockpit-oauth-client"
                    },
                    {
                        "name": "KUBERNETES_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "COCKPIT_KUBE_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "REGISTRY_ONLY",
                        "value": "true"
                    },
                    {
                        "name": "REGISTRY_HOST",
                        "value": "docker-registry-default.apps.claytontest.origin-ci-int-gce.dev.rhcloud.com"
                    }
                ],
                "image": "cockpit/kubernetes:latest",
                "imagePullPolicy": "Always",
                "livenessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "initialDelaySeconds": 10,
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "name": "registry-console",
                "ports": [
                    {
                        "containerPort": 9090,
                        "protocol": "TCP"
                    }
                ],
                "readinessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "resources": {},
                "securityContext": {
                    "capabilities": {
                        "drop": [
                            "KILL",
                            "MKNOD",
                            "SETGID",
                            "SETUID"
                        ]
                    },
                    "privileged": false,
                    "runAsUser": 1000000000,
                    "seLinuxOptions": {
                        "level": "s0:c1,c0"
                    }
                },
                "terminationMessagePath": "/dev/termination-log",
                "terminationMessagePolicy": "File",
                "volumeMounts": [
                    {
                        "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                        "name": "default-token-r7r5f",
                        "readOnly": true
                    }
                ]
            }
        ],
        "dnsPolicy": "ClusterFirst",
        "imagePullSecrets": [
            {
                "name": "default-dockercfg-5czxp"
            }
        ],
        "nodeName": "ci-claytontest-ig-n-63c4",
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": {
            "fsGroup": 1000000000,
            "seLinuxOptions": {
                "level": "s0:c1,c0"
            }
        },
        "serviceAccount": "default",
        "serviceAccountName": "default",
        "terminationGracePeriodSeconds": 30,
        "volumes": [
            {
                "name": "default-token-r7r5f",
                "secret": {
                    "defaultMode": 420,
                    "secretName": "default-token-r7r5f"
                }
            }
        ]
    },
    "status": {
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "Initialized"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:43:05Z",
                "status": "True",
                "type": "Ready"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "PodScheduled"
            }
        ],
        "containerStatuses": [
            {
                "containerID": "docker://ef5aef49b57a811f642bba553a6a1c4dde865289c825234527e1c2b3aeceea12",
                "image": "docker.io/cockpit/kubernetes:latest",
                "imageID": "docker-pullable://docker.io/cockpit/kubernetes@sha256:a4b418f80f5654021ff63526bbb9644f23a5cc56b6eb50d684f21c1aaf324256",
                "lastState": {},
                "name": "registry-console",
                "ready": true,
                "restartCount": 0,
                "state": {
                    "running": {
                        "startedAt": "2017-11-03T23:43:00Z"
                    }
                }
            }
        ],
        "hostIP": "10.142.0.2",
        "phase": "Running",
        "podIP": "172.16.0.4",
        "qosClass": "BestEffort",
        "startTime": "2017-11-03T23:42:58Z"
    }
}

# pod UPDATE event - readiness changes, but not status info like transition time
{
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": {
        "annotations": {
            "kubernetes.io/created-by": "{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"registry-console-1\",\"uid\":\"4b81a94b-c0eb-11e7-b34d-42010a8e0004\",\"apiVersion\":\"v1\",\"resourceVersion\":\"5599\"}}\n",
            "openshift.io/deployment-config.latest-version": "1",
            "openshift.io/deployment-config.name": "registry-console",
            "openshift.io/deployment.name": "registry-console-1",
            "openshift.io/generated-by": "OpenShiftNewApp",
            "openshift.io/scc": "restricted"
        },
        "creationTimestamp": "2017-11-03T23:42:58Z",
        "generateName": "registry-console-1-",
        "labels": {
            "app": "registry-console",
            "deployment": "registry-console-1",
            "deploymentconfig": "registry-console",
            "name": "registry-console"
        },
        "name": "registry-console-1-lc9g6",
        "namespace": "default",
        "ownerReferences": [
            {
                "apiVersion": "v1",
                "blockOwnerDeletion": true,
                "controller": true,
                "kind": "ReplicationController",
                "name": "registry-console-1",
                "uid": "4b81a94b-c0eb-11e7-b34d-42010a8e0004"
            }
        ],
        "resourceVersion": "7000",
        "selfLink": "/api/v1/namespaces/default/pods/registry-console-1-lc9g6",
        "uid": "b8e9cdc0-c0f0-11e7-b34d-42010a8e0004"
    },
    "spec": {
        "containers": [
            {
                "env": [
                    {
                        "name": "OPENSHIFT_OAUTH_PROVIDER_URL",
                        "value": "https://api.claytontest.origin-ci-int-gce.dev.rhcloud.com/"
                    },
                    {
                        "name": "OPENSHIFT_OAUTH_CLIENT_ID",
                        "value": "cockpit-oauth-client"
                    },
                    {
                        "name": "KUBERNETES_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "COCKPIT_KUBE_INSECURE",
                        "value": "false"
                    },
                    {
                        "name": "REGISTRY_ONLY",
                        "value": "true"
                    },
                    {
                        "name": "REGISTRY_HOST",
                        "value": "docker-registry-default.apps.claytontest.origin-ci-int-gce.dev.rhcloud.com"
                    }
                ],
                "image": "cockpit/kubernetes:latest",
                "imagePullPolicy": "Always",
                "livenessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "initialDelaySeconds": 10,
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "name": "registry-console",
                "ports": [
                    {
                        "containerPort": 9090,
                        "protocol": "TCP"
                    }
                ],
                "readinessProbe": {
                    "failureThreshold": 3,
                    "httpGet": {
                        "path": "/ping",
                        "port": 9090,
                        "scheme": "HTTP"
                    },
                    "periodSeconds": 10,
                    "successThreshold": 1,
                    "timeoutSeconds": 5
                },
                "resources": {},
                "securityContext": {
                    "capabilities": {
                        "drop": [
                            "KILL",
                            "MKNOD",
                            "SETGID",
                            "SETUID"
                        ]
                    },
                    "privileged": false,
                    "runAsUser": 1000000000,
                    "seLinuxOptions": {
                        "level": "s0:c1,c0"
                    }
                },
                "terminationMessagePath": "/dev/termination-log",
                "terminationMessagePolicy": "File",
                "volumeMounts": [
                    {
                        "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                        "name": "default-token-r7r5f",
                        "readOnly": true
                    }
                ]
            }
        ],
        "dnsPolicy": "ClusterFirst",
        "imagePullSecrets": [
            {
                "name": "default-dockercfg-5czxp"
            }
        ],
        "nodeName": "ci-claytontest-ig-n-63c4",
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": {
            "fsGroup": 1000000000,
            "seLinuxOptions": {
                "level": "s0:c1,c0"
            }
        },
        "serviceAccount": "default",
        "serviceAccountName": "default",
        "terminationGracePeriodSeconds": 30,
        "volumes": [
            {
                "name": "default-token-r7r5f",
                "secret": {
                    "defaultMode": 420,
                    "secretName": "default-token-r7r5f"
                }
            }
        ]
    },
    "status": {
        "conditions": [
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "Initialized"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:43:05Z",
                "status": "False",
                "type": "Ready"
            },
            {
                "lastProbeTime": null,
                "lastTransitionTime": "2017-11-03T23:42:58Z",
                "status": "True",
                "type": "PodScheduled"
            }
        ],
        "containerStatuses": [
            {
                "containerID": "docker://ef5aef49b57a811f642bba553a6a1c4dde865289c825234527e1c2b3aeceea12",
                "image": "docker.io/cockpit/kubernetes:latest",
                "imageID": "docker-pullable://docker.io/cockpit/kubernetes@sha256:a4b418f80f5654021ff63526bbb9644f23a5cc56b6eb50d684f21c1aaf324256",
                "lastState": {},
                "name": "registry-console",
                "ready": true,
                "restartCount": 0,
                "state": {
                    "running": {
                        "startedAt": "2017-11-03T23:43:00Z"
                    }
                }
            }
        ],
        "hostIP": "10.142.0.2",
        "phase": "Running",
        "podIP": "172.16.0.4",
        "qosClass": "BestEffort",
        "startTime": "2017-11-03T23:42:58Z"
    }
}
```

There's another bug here - it looks like the pod status was also not updated.  Going to spawn an upstream.

**Comment by smarterclayton** [comment]
I think the bug in openshift-sdn is that the hostsubnet changed, but we still have pods running on the node that are in the old subnet, and they are not restarted to get new IPs (probably due to a race, the network plugin sees the old subnet and hands it out).

We need to ensure that:

1. openshift-sdn is ensuring the most recent subnet assignment is in effect (why isn't that happening here?)
2. if possible, we correlate UIDs between hostsubnet and node object in openshift-sdn (do we have this data) and ignore hostsubnets that don't match the node

Because this will always be racy when it happens fast (which is what we want, since we want new nodes up asap), some racing is inevitable.  We just need to be able to reconcile and correct the problem when detected.

**Comment by danwinship** [comment]
> I think the bug in openshift-sdn is that the hostsubnet changed, but we still have pods running on the node that are in the old subnet, and they are not restarted to get new IPs (probably due to a race, the network plugin sees the old subnet and hands it out).

It's not even a race. The SDN code currently doesn't even consider the possibility that the node's HostSubnet might have changed out from under existing pods.

> if possible, we correlate UIDs between hostsubnet and node object in openshift-sdn (do we have this data) and ignore hostsubnets that don't match the node

Vaguely related: if the SDN controller misses a Node DELETED event then it will never clean up the corresponding HostSubnet. https://bugzilla.redhat.com/show_bug.cgi?id=1470612#c4

**Comment by smarterclayton** [comment]
Ok, so the easy thing is to hook host subnet changing into the existing
health check behavior (for OVS) and have the process restart itself if it
detects a change.  In the future, if we can restore without restart,
that'll be better.

**Comment by kincl** [comment]
This also happens when switching to a cloud plugin like vSphere with an existing cluster. I also experienced this when I cherry-picked https://github.com/kubernetes/kubernetes/issues/59602 which changes the vSphere node externalID and I had to restart docker daemons across the cluster so when Origin rebases 1.9 or upgrades to 1.10 vSphere people running the openshift-sdn will be hit by this again.

I did notice that some pods did restart automatically if they were using a liveness check that was network-based (HTTP/TCP) because connectivity to the pod from the kubelet was lost.
